mailbox/event/event-memory/src/main/java/org/apache/james/mailbox/events/delivery/EventDelivery.java

```diff
@@ -17,36 +17,36 @@
  * under the License.                                           *
  ****************************************************************/
 
-package org.apache.james.mailbox.store.event;
+package org.apache.james.mailbox.events.delivery;
 
-import javax.annotation.PreDestroy;
+import java.util.Collection;
 
 import org.apache.james.mailbox.Event;
 import org.apache.james.mailbox.MailboxListener;
 
-public class MixedEventDelivery implements EventDelivery {
+import reactor.core.publisher.Mono;
 
-    private final AsynchronousEventDelivery asynchronousEventDelivery;
-    private final SynchronousEventDelivery synchronousEventDelivery;
+public interface EventDelivery {
+    class ExecutionStages {
+        private final Mono<Void> synchronousListenerFuture;
+        private final Mono<Void> asynchronousListenerFuture;
 
-    public MixedEventDelivery(AsynchronousEventDelivery asynchronousEventDelivery,
-                              SynchronousEventDelivery synchronousEventDelivery) {
-        this.asynchronousEventDelivery = asynchronousEventDelivery;
-        this.synchronousEventDelivery = synchronousEventDelivery;
-    }
+        ExecutionStages(Mono<Void> synchronousListenerFuture, Mono<Void> asynchronousListenerFuture) {
+            this.synchronousListenerFuture = synchronousListenerFuture;
+            this.asynchronousListenerFuture = asynchronousListenerFuture;
+        }
 
-    @Override
-    public void deliver(MailboxListener mailboxListener, Event event) {
-        if (mailboxListener.getExecutionMode().equals(MailboxListener.ExecutionMode.SYNCHRONOUS)) {
-            synchronousEventDelivery.deliver(mailboxListener, event);
-        } else {
-            asynchronousEventDelivery.deliver(mailboxListener, event);
+        public Mono<Void> synchronousListenerFuture() {
+            return synchronousListenerFuture;
         }
-    }
 
-    @PreDestroy
-    public void stop() {
-        asynchronousEventDelivery.stop();
+        public Mono<Void> allListenerFuture() {
```

Comment:
I'm just wondering where is this method is suppose to used in our code (not test) ?

Reply:
It is tested. See InVmEventDeliveryTest.java `deliverShouldHaveCalledSynchronousListenersWhenAllListenerExecutedJoined` & `deliverShouldHaveCalledAsynchronousListenersWhenAllListenerExecutedJoined`

The idea is to keep the capabilities to await all listener (CF SynchronousEventDelivery), await none (AsyncEventDelivery) or just some of them (MixedEventDelivery) with a single implem.

Then choosing the execution model is a caller choice and not an implementation detail.